### DISCUSSION TITLE
esp32: gpio: blinky: Open-drain outputs not working on ESP32-CAM

### DISCUSSION THREAD

**Discussion by towen** [zephyrproject-rtos/zephyr#65244] [Q&A] [answered]
Not sure if this is a real zephyr issue, a hardware issue with my cheapo board, or a simple PEBKAC, so posting a discussion.
I have an ESP32-CAM module ([schematic here](https://i0.wp.com/randomnerdtutorials.com/wp-content/uploads/2020/03/ESP32-CAM-AI-Thinker-schematic-diagram.png)) and I am unable to blink the LED. The LED turns on when I call `gpio_pin_configure_dt`, but it will not turn off after that.

The board is very similar to the `esp32_devkitc_wroom`, so I have made a minimal DTS overlay for use with the 'blinky' sample: `blinky/boards/esp32_devkitc_wroom.overlay`
```devicetree
/ {

   leds {
       compatible="gpio-leds";
       led_r: led_r {
                        gpios = <&gpio0 33 (GPIO_PULL_UP | GPIO_ACTIVE_LOW)>;
                        label = "Red LED 1";
       };
       led_w: led_w {
                        gpios = <&gpio0 4 (GPIO_PULL_DOWN | GPIO_ACTIVE_HIGH)>;
                        label = "Camera Flash LED (Bright White)";
       };
   };

   aliases {
         led0 = &led_r;
         led1 = &led_w;
   };

};

```

The red LED (which is configured as open-drain, one side connected to VCC) turns on, but does not blink. 
Whereas the white LED (which uses a NPN transistor so is active high) works correctly.  
I have tried GPIO_PUSH_PULL, and I have tried setting ACTIVE_HIGH but it still does not blink (instead it remains off) and I have also tried the LED shell, which also cannot change the LED from its initial state.

There is a second open-drain GPIO on this board, which supplies the camera (via a P-FET), and it would appear that the camera supply also will not turn on (I see no voltage on pin 10 of the camera connector when I try to turn it on, and all i2c requests result in Error -14 -EFAULT). 
I think that could be the issue behind [this unanswered stackoverflow](https://stackoverflow.com/questions/75481551/cannot-communicate-with-ov2640-sensor-on-esp32cam-using-zephyros) from someone struggling to use the same board with Zephyr..

(I am using zephyr-v3.5.0-1182-g26002b060708 i.e. 'main' as of a few days ago)

**Comment by ndrs-pst** [comment]
I think you need to set extra flag as (GPIO_OUTPUT | GPIO_OPEN_DRAIN) when setup the pin by this function 
`int gpio_pin_configure_dt(const struct gpio_dt_spec* spec, gpio_flags_t extra_flags)`

Also please refer to underly driver function gpio_esp32_config in gpio_esp32.c for more information.

**Reply by towen** [reply]
Hi, thanks. Unfortunately setting extra flags as `(GPIO_OUTPUT |  GPIO_OPEN_DRAIN)` in main.c, and/or adding GPIO_OPEN_DRAIN to the devicetree (which I had tried earlier but forgot to mention), does not make any difference.  The LED turns on but doesn't blink. 

EDIT: OK, so if I add an open-drain LED onto GPIO13 of my board and modify the devicetree to use GPIO13, then it blinks. 
So basically I don't understand why GPIO33 is behaving differently to GPIO13.

**Comment by ndrs-pst** [comment, marked as answer]
After I take a look at gpio_esp32.c (gpio_esp32_config) 
May be the problem come from which only can check up to (0 ~ 31)
```
static inline bool gpio_pin_is_valid(uint32_t pin) {
    return ((BIT(pin) & SOC_GPIO_VALID_GPIO_MASK) != 0);
}
```
Let's see if you change BIT(pin) -> BIT64(pin) instead.

**Sorry, after checking with reference manual you should define GPIO33 as &gpio1 1 instead.**
